File: app/pricing.py
"""Money math for orders.

Prices are floats at the API edge; every calculation here goes through
Decimal so half-cent values round predictably, not the way binary floats do.
Each line is rounded to the cent before the lines are summed, which is how the
line items on a printed invoice add up.

Half-cent values use banker's rounding (half to even), which is the usual
default in financial software because it does not bias a long run of sums
upward the way half-up does.
"""

from decimal import ROUND_HALF_EVEN, Decimal
from typing import Protocol

CENT = Decimal("0.01")


class Line(Protocol):
    quantity: int
    unit_price: float


def line_total(quantity: int, unit_price: float) -> Decimal:
    """One line, rounded to the cent."""
    amount = Decimal(str(unit_price)) * quantity
    return amount.quantize(CENT, rounding=ROUND_HALF_EVEN)


def order_total(items: list[Line]) -> float:
    """Sum of the rounded line totals, as the float the API returns."""
    total = sum((line_total(item.quantity, item.unit_price) for item in items), Decimal("0"))
    return float(total.quantize(CENT, rounding=ROUND_HALF_EVEN))

File: services/orders-node/src/pricing.mjs
// Money math for orders. Every line is rounded to the cent, half up, before
// the lines are summed, which is how the line items on a printed invoice add
// up. Integer cents inside; floats only at the API edge.
//
// Half-up on the line's exact decimal, not on its binary float: q * unitPrice
// in cents is computed on the float product and rounded half away from zero,
// which matches the other services for every price with at most two decimals
// and agrees with them on the half-cent boundary cases.

export function lineTotalCents(quantity, unitPrice) {
  return Math.round(quantity * unitPrice * 100);
}

export function orderTotal(items) {
  const cents = items.reduce((sum, item) => sum + lineTotalCents(item.quantity, item.unitPrice), 0);
  return cents / 100;
}
